### PR TITLE
[feature] Add JS testing framework and a few initial tests

### PR DESCRIPTION
**Pull request by sloria**
Adds an automated JS testing harness, using the following libraries:

- Karma for running browser tests
- Mocha test framework
- Chai asserts
- Sinon for mocks, spies, fake server

Also adds a few unit tests for `osfHelpers.js`, `profile.js`, and `dropboxNodeConfig.js`. Coverage is far from comprehensive; these are more of a starting point for future testing.

Todo:

- [x] PhantomJS support
- [x] Firefox and Chrome support
- [x] Travis CI
- [x] Add example async test
- [x] Add example KO ViewModel test
- [x] Add example addon tests
- [ ] Update dev docs (post-merge)
- [x] ES6 or Coffeescript for less verbose tests?
- [ ] SauceLabs support? (probably post-merge)

Side effects

Minor changes to the application code were made to make testing possible. For example, I exposed one of the Knockout viewmodels in `profile.js`: https://github.com/CenterForOpenScience/osf.io/blob/5b3e228aa650c133191e3e4219eebf9e0a4e9509/website/static/js/profile.js#L775-776 and added callbacks to methods that made an AJAX request (example: https://github.com/CenterForOpenScience/osf.io/blob/5b3e228aa650c133191e3e4219eebf9e0a4e9509/website/static/js/profile.js#L277-287). 

Devs will need to install the new requirements by running `npm install`.

### CODE REVIEW

**Comment by sloria** [comment]
This is ready for review

**Comment by lbanner** [comment]
@sloria -- who do you suggest to review?

**Comment by brianjgeiger** [comment]
Our linter does not like this format at *all*.

screen shot 2015-03-09 at 10 49 01 am

**Comment by sloria** [comment]
@brianjgeiger If your IDE supports jshint, use that, as we have configured it to handle ES6 syntax: https://github.com/CenterForOpenScience/osf.io/blob/5b3e228aa650c133191e3e4219eebf9e0a4e9509/.jshintrc#L17

**Comment by brianjgeiger** [comment]
Hmm, it is being stubborn, but I'll see if I can get to to work properly.

**Comment by brianjgeiger** [comment]
It just took a settings change and a reboot to get that fixed. Much better.

**Comment by brianjgeiger** [comment]
Synchronous unit tests: √

screen shot 2015-03-09 at 1 11 34 pm

File: website/static/js/tests/osfHelpers.test.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    });
+
+    describe('handleJSONError', () => {
+        var response = {
+            responseJSON: {message_short: 'Oh no!', message_long: 'Something went wrong'}
+        };
+        it('uses the response body if available', () => {
+            var spy = new sinon.spy($osf, 'growl');
+            $osf.handleJSONError(response);
+            assert.called(spy);
```
Comment (brianjgeiger) [inline website/static/js/tests/osfHelpers.test.js]: Does this actually test that it's using the response body if available? Or just that the growlbox was called when you handle a json error?

Comment (sloria) [inline website/static/js/tests/osfHelpers.test.js]: You're right. I meant to put a TODO there.

Comment (sloria) [inline website/static/js/tests/osfHelpers.test.js]: Finished that up in  acb32f6

File: website/addons/dropbox/static/tests/dropboxNodeConfig.test.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        server = utils.createServer(sinon, endpoints);
+    });
+
+    after(() => {
+        server.restore();
+    });
+
+    describe('ViewModel', () => {
+        it('fetches data from the server on initialization', (done) => {
+            var vm = new dropboxNodeConfig._ViewModel('/api/v1/12345/dropbox/config/', '', '', function() {
```
Comment (icereval) [inline website/addons/dropbox/static/tests/dropboxNodeConfig.test.js]: could this return a (deferred) promise?

Comment (sloria) [inline website/addons/dropbox/static/tests/dropboxNodeConfig.test.js]: Would the ViewModel constructor return a promise? What would it resolve to?

Comment (sloria) [inline website/addons/dropbox/static/tests/dropboxNodeConfig.test.js]: Also, we need to call `server.respond()` after constructing the `ViewModel`, so it wouldn't be feasible to `return` before that.

However, I just found that sinon's `fakeServer` has an `autoRespond` option, which will call `server.respond()` after request. I'll try that

**Comment by caneruguz** [comment]
I'm not in favor of adding Mocha and ES6 at the same time in the same change. It makes the learning process more difficult than necessary, especially if we want everyone to start writing tests for js, suddenly the syntax will look different enough to be confusing.

**Comment by jmcarp** [comment]
NBD, but it looks like this setup would publicly serve test files given our current nginx configuration. We should probably only be serving the `public` path(s) at this point.

**Comment by sloria** [comment]
+1 @jmcarp , though we'd have to be careful, as we are loading some files from `/static/vendor/bower_components` at the moment.

**Pull request merged**